### DISCUSSION TITLE
Heatmap plotting variable deviation or Z- scores of TF motifs

### DISCUSSION THREAD

**Discussion by JimeTosello** [GreenleafLab/ArchR#698] [General]
Hello, 

We have tried to do a heatmap plotting variable deviation or Z- Scores of TF motifs. We have applied getMarkerFeatures() using useSeqnames to select "z" or "deviation". This worked, but when we tried plotMarkerHeatmap(), it didn't worked. We noticed that this function generate before a heatmap of clusters but not at single nucleus as the one from peak2gene plot for example. 
Anyway, we obtained the matrices so we can try outside of ArchR, but we are sure it is possible to get it from ArchR. Sorry if we am missing an existing function or path!

Thanks for ArchR! It is a fantastic tool ♡ !

______

markerMotifs <- getFeatures(peakCalls_Matrix, useMatrix = "MotifMatrix")
markerMotifs <- grep("z:", markerMotifs, value = TRUE)

markersMM.slctRes <- getMarkerFeatures(
    ArchRProj = peakCalls_Matrix, 
    useMatrix = "MotifMatrix", 
    groupBy = paste("Clusters",slct.res,sep="."),
    #bias = c("TSSEnrichment", "log10(nFrags)"),
    useSeqnames="z",
    testMethod = "wilcoxon"
)
 
plotMarkerHeatmap(
  seMarker = markersMM.slctRes,
   transpose = FALSE
)
 ArchR logging to : ArchRLogs/ArchR-plotMarkerHeatmap-2489c485ee347-Date-2021-04-16_Time-19-27-34.log
If there is an issue, please report to github with logFile!
Error in bS[[1]][, colnames(mat), drop = FALSE] : indice hors limites
De plus : Warning message:
In plotMarkerHeatmap(seMarker = markersMM.slctRes, returnMatrix = FALSE,  :
  production de NaN

**Comment by cswoboda** [comment]
Hey, did you perhaps try to subset the project at any point to generate these matrices? There's a pretty common bug that if you subset prior to an analysis plotMarkerHeatmap won't work as NAs were introduced

**Reply by JimeTosello** [reply]
Hello, thanks for the answer. We have checked and we haven't done a subset before. We explored and the NA appeared after the getMarkerFeatures(). We don't complete understand what we are doing wrong, but we used getMatrixFromProject() to obtain the deviations and z matrix, and we verify that they do not contain NA. It is only after the getMarkerFeatures() that the NA appeared in the matrix. We tried to remove all the filters for getMarkerFeatures() function, but the NA are still there. We can send you the log files of the 2 functions if you prefer. Sorry if we are missing something! 
____

markersMM.slctRes <- getMarkerFeatures(
     ArchRProj = peakCalls_Matrix, 
     useMatrix = "MotifMatrix", 
     groupBy = paste("Clusters",slct.res,sep="."),
     bias = c("TSSEnrichment", "log10(nFrags)"),
     useSeqnames="z",
     testMethod = "wilcoxon"
 )
ArchR logging to : ArchRLogs/ArchR-getMarkerFeatures-2d0d9401a4a7d-Date-2021-04-20_Time-12-30-38.log
If there is an issue, please report to github with logFile!
2021-04-20 12:30:38 : Matching Known Biases, 0.001 mins elapsed.
###########
2021-04-20 12:30:50 : Completed Pairwise Tests, 0.203 mins elapsed.
###########
ArchR logging successful to : ArchRLogs/ArchR-getMarkerFeatures-2d0d9401a4a7d-Date-2021-04-20_Time-12-30-38.log
____
heatmapGSmatrix <- plotMarkerHeatmap(
     seMarker = markersMM.slctRes, 
     returnMatrix = FALSE,
     transpose = FALSE
 )
ArchR logging to : ArchRLogs/ArchR-plotMarkerHeatmap-2d0d91e8c2fe8-Date-2021-04-20_Time-12-31-53.log
If there is an issue, please report to github with logFile!
Error in bS[[1]][, colnames(mat), drop = FALSE] : indice hors limites
De plus : Warning message:
In plotMarkerHeatmap(seMarker = markersMM.slctRes, returnMatrix = FALSE,  :
  production de NaN

**Comment by rcorces** [comment]
This may be better handled as an issue report. Please check the output getMarkerFeatures and compare it to what it should look like based on the tutorial data. If you cannot figure out your problem, post it as an issue and include a detailed report of what you have tried and include the log file.

**Comment by Cailu086Lin** [comment]
Hello,

Is there a way to export data (See attached plot)? I want to plot it in other way.

Many thanks,

heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markersPeaks, 
  cutOff = "FDR <= 0.01 & Log2FC >= 0.5",
  transpose = TRUE)
 
 
[Fig.pdf](https://github.com/GreenleafLab/ArchR/files/7172633/Fig.pdf)

**Reply by rcorces** [reply]
see `returnMatrix` https://www.archrproject.com/reference/plotMarkerHeatmap.html
